R/surv-roc_survival_curve.R

```diff
@@ -0,0 +1,169 @@
+#' columns `.threshold`, `recall`, and `precision`.
+#'
+#' @seealso
+#' Compute the area under the ROC survival curve with (TODO link to
+#' roc_survival_auc()).
+#'
+#' @author Emil Hvitfeldt
+#' @examples
+#' roc_survival_curve(
+#'   lung_surv,
+#'   truth = surv_obj,
+#'   estimate = .pred_survival,
+#'   censoring_weights = ipcw,
+#'   eval_time = .time
+#' )
+#' @export
+roc_survival_curve <- function(data, ...) {
+  UseMethod("roc_survival_curve")
+}
+
+#' @export
+#' @rdname roc_survival_curve
+roc_survival_curve.data.frame <- function(data,
+                                          truth,
+                                          estimate,
+                                          censoring_weights,
+                                          eval_time,
+                                          na_rm = TRUE,
+                                          case_weights = NULL,
+                                          ...) {
+
+  result <- curve_survival_metric_summarizer(
+    name = "roc_survival_curve",
+    fn = roc_survival_curve_vec,
+    data = data,
+    truth = !!enquo(truth),
+    estimate = !!enquo(estimate),
+    censoring_weights = !!enquo(censoring_weights),
+    eval_time = !!enquo(eval_time),
+    na_rm = na_rm,
+    case_weights = !!enquo(case_weights)
+  )
+
+  curve_finalize(result, data, "roc_survival_df", "grouped_roc_survival_df")
+}
+
+roc_survival_curve_vec <- function(truth,
+                                   estimate,
+                                   censoring_weights,
+                                   eval_time,
+                                   na_rm = TRUE,
+                                   case_weights = NULL,
+                                   ...) {
+  check_dynamic_survival_metric(
+    truth, estimate, censoring_weights, case_weights, eval_time
+  )
+
+  if (na_rm) {
+    result <- yardstick_remove_missing(
+      truth, estimate, case_weights, censoring_weights, eval_time
+    )
+
+    truth <- result$truth
+    estimate <- result$estimate
+    censoring_weights <- result$censoring_weights
+    eval_time <- result$eval_time
+    case_weights <- result$case_weights
+  } else {
+    any_missing <- yardstick_any_missing(
+      truth, estimate, case_weights, censoring_weights, eval_time
+    )
+    if (any_missing) {
+      return(NA_real_)
+    }
+  }
+
+  roc_survival_curve_impl(
+    truth = truth,
+    estimate = estimate,
+    censoring_weights = censoring_weights,
+    eval_time = eval_time
+  )
+}
+
+roc_survival_curve_impl <- function(truth,
+                                    estimate,
+                                    censoring_weights,
+                                    eval_time) {
+  res <- dplyr::tibble(.threshold = sort(unique(c(0, 1, estimate))))
+  res$sensitivity <- vapply(
+    res$.threshold,
+    sensitivity_uno_2007,
+    FUN.VALUE = numeric(1),
+    eval_time, truth, estimate, censoring_weights
+  )
+  res$specificity <- vapply(
```

Comment:
It would also be good to have someone review this for efficiency. The functions for sensitivity and specificity were working prototypes and it would be good to know if the `vapply()` strategy over thresholds is efficient or not.

Maybe that is another PR but I think that these could be made to run faster.

Reply:
Added an issue to do it in a separate PR here https://github.com/tidymodels/yardstick/issues/379